### ISSUE TITLE
Catboost with user defined objective cannot compute ShapValues

### ISSUE THREAD

**Issue by antipisa** [catboost/catboost#2377]
Problem: get_feature_importance(data=X, type="ShapValues") when model has been trained with custom objective
catboost version: 1.1.1
Operating System: Linux
CPU: Yes
GPU: No

Computing the ShapValues of a CatboostRegressor model that has already been fit with a custom objective function, and passing in a new pool X triggers the `CatboostError: "User defined objectives/metrics require Target data"`. Hence one cannot call 
`
bst.get_feature_importance(data=X, type="ShapValues")
`
without passing in label. Is there a way around this? Can we call _calc_fstr with one of the standard loss functions instead for the purpose of computing shap?

**Comment by ek-ak** [comment]
Hello! Thank you for this bug report!
I was able to reproduce `"User defined objective/metrics require Target data" `error by calling`model.get_feature_importance(data, reference_data=ref_data, type="ShapValues")` only, not by calling `model.get_feature_importance(data, type="ShapValues")`. 
For example next code doesn't fail:
```
class RmseObjective(object):
    def calc_ders_range(self, approxes, targets, weights):
        assert len(approxes) == len(targets)
        if weights is not None:
            assert len(weights) == len(approxes)

        result = []
        for index in range(len(targets)):
            der1 = targets[index] - approxes[index]
            der2 = -1

            if weights is not None:
                der1 *= weights[index]
                der2 *= weights[index]

            result.append((der1, der2))
        return result

class RmseMetric(object):
    def get_final_error(self, error, weight):
        return np.sqrt(error / (weight + 1e-38))

    def is_max_optimal(self):
        return False

    def evaluate(self, approxes, target, weight):
        assert len(approxes) == 1
        assert len(target) == len(approxes[0])

        approx = approxes[0]

        error_sum = 0.0
        weight_sum = 0.0

        for i in range(len(approx)):
            w = 1.0 if weight is None else weight[i]
            weight_sum += w
            error_sum += w * ((approx[i] - target[i])**2)

        return error_sum, weight_sum

model = CatBoostRegressor(
    iterations=10,
    random_seed=42, 
    loss_function=RmseObjective(), 
    eval_metric=RmseMetric()
)

model.fit(Pool(X, y_train))

data = Pool(X)
model.get_feature_importance(data, type="ShapValues")
```

Could you give a minimal reproducible example, please?

**Comment by antipisa** [comment]
```
class RmseObjective(object):
    def calc_ders_range(self, approxes, targets, weights, alpha=0.9):
        assert len(approxes) == len(targets)
        if weights is not None:
            assert len(weights) == len(approxes)

        result = []
        for index in range(len(targets)):
            der1 = targets[index] - (1+alpha)*approxes[index]
            der2 = -(1+alpha)

            if weights is not None:
                der1 *= weights[index]
                der2 *= weights[index] + alpha

            result.append((der1, der2))
        return result
    
class RmseMetric(object):
    def get_final_error(self, error, weight):
        return np.sqrt(error / (weight + 1e-38))

    def is_max_optimal(self):
        return False

    def evaluate(self, approxes, target, weight):
        assert len(approxes) == 1
        assert len(target) == len(approxes[0])

        approx = approxes[0]

        error_sum = 0.0
        weight_sum = 0.0

        for i in range(len(approx)):
            w = 1.0 if weight is None else weight[i]
            weight_sum += w
            error_sum += w * ((approx[i] - target[i])**2)

        return error_sum, weight_sum
    
from pathlib import Path
Path('testmodels').mkdir(parents=True, exist_ok=True)


N, p = 1000, 10
Ntest = 500

params = dict(num_boost_round=500, learning_rate=0.05, loss_function=RmseObjective(), eval_metric='R2:use_weights=True', custom_metric='RMSE:use_weights=True', boosting_type='Ordered', border_count=10, feature_border_type='UniformAndQuantiles', model_size_reg=0, l2_leaf_reg=2, max_depth=8, score_function='Cosine', subsample=0.8, penalties_coefficient=1, leaf_estimation_method='Newton', grow_policy='SymmetricTree', ctr_leaf_count_limit=None, has_time=False, fold_permutation_block=100, metric_period=1, colsample_bylevel=1, sampling_frequency='PerTreeLevel', bootstrap_type='MVS', random_strength=0, early_stopping_rounds=10, random_seed=3, thread_count=20, allow_writing_files=False)
model = CatBoostRegressor(**params)


def create_data(N, p):
    X = np.random.randn(N, p)
    w = np.abs(np.random.randn(N))
    # create 2 categorical features
    catarr1 = np.random.randint(3, size=(N,))
    catarr2 = np.random.randint(10, size=(N,))
    X = pd.DataFrame(np.column_stack((X, catarr1, catarr2)))
    y = np.random.randn(N)
    X.loc[:,p] = X.loc[:,p].astype(str)
    X.loc[:,p+1] = X.loc[:,p+1].astype(str)
    return X, y, w

X, y, w = create_data(N, p)
feature_names = list(np.arange(p+2))

trainpool = Pool(data=X, label=y, cat_features=[p, p+1], weight=w, feature_names=feature_names)
pool_no_label = cb.Pool(data=X, label=None,  cat_features=[p, p+1],  feature_names=feature_names)

model.fit(trainpool, eval_set=None, verbose_eval=500, plot=True)

shapvals = model.get_feature_importance(data=pool_no_label, prettified=True, type='ShapValues', reference_data=None)
explainer = shap.TreeExplainer(model)
shap_values = explainer.shap_values(pool_no_label)

filename = "testmodels/custom_bst"
model.save_model(filename, format='cbm', pool=trainpool)

model2 = CatBoostRegressor().load_model(filename)

X_test, y_test, w_test = create_data(Ntest, p)
testpool = cb.Pool(data=X_test, label=None,  cat_features=[p, p+1],  feature_names=feature_names)

shapvals2 = model2.get_feature_importance(data=testpool, prettified=True, type='ShapValues', reference_data=None)

shap.summary_plot(shapvals2.values[:,:-1], features=X_test)
```

**Comment by antipisa** [comment]
Python 3.10
Catboost 1.1

```

def MyLossFunction():
    return RmseObjective()
    

class RmseObjective(object):
    def calc_ders_range(self, approxes, targets, weights, alpha=1.1):
        assert len(approxes) == len(targets)
        if weights is not None:
            assert len(weights) == len(approxes)

        result = []
        for index in range(len(targets)):
            der1 = targets[index] - (1+alpha)*approxes[index]
            der2 = -(1+alpha)

            if weights is not None:
                der1 *= weights[index]
                der2 *= weights[index] + alpha

            result.append((der1, der2))
        return result
    
class RmseMetric(object):
    def get_final_error(self, error, weight):
        return np.sqrt(error / (weight + 1e-38))

    def is_max_optimal(self):
        return False

    def evaluate(self, approxes, target, weight):
        assert len(approxes) == 1
        assert len(target) == len(approxes[0])

        approx = approxes[0]

        error_sum = 0.0
        weight_sum = 0.0

        for i in range(len(approx)):
            w = 1.0 if weight is None else weight[i]
            weight_sum += w
            error_sum += w * ((approx[i] - target[i])**2)

        return error_sum, weight_sum
    
from pathlib import Path
Path('testmodels').mkdir(parents=True, exist_ok=True)


N, p = 1000, 10
Ntest = 500

params = dict(num_boost_round=1000, learning_rate=0.05, loss_function=MyLossFunction(), eval_metric='R2:use_weights=True', custom_metric='RMSE:use_weights=True', boosting_type='Ordered', border_count=20, feature_border_type='UniformAndQuantiles', model_size_reg=0, l2_leaf_reg=2, max_depth=8, score_function='Cosine', subsample=0.8, penalties_coefficient=1, leaf_estimation_method='Newton', grow_policy='SymmetricTree', ctr_leaf_count_limit=None, has_time=False, fold_permutation_block=100, metric_period=1, colsample_bylevel=1, sampling_frequency='PerTreeLevel', bootstrap_type='MVS', random_strength=0, early_stopping_rounds=10, random_seed=3, thread_count=20, allow_writing_files=False, train_dir='var/tmp/usr')
model = CatBoostRegressor(**params)


def create_data(N, p):
    X = np.random.randn(N, p)
    w = np.abs(np.random.randn(N))
    # create 2 categorical features
    catarr1 = np.random.randint(3, size=(N,))
    catarr2 = np.random.randint(10, size=(N,))
    X = pd.DataFrame(np.column_stack((X, catarr1, catarr2)))
    y = np.random.randn(N)
    X.loc[:,p] = X.loc[:,p].astype(str)
    X.loc[:,p+1] = X.loc[:,p+1].astype(str)
    return X, y, w

X, y, w = create_data(N, p)
feature_names = list(np.arange(p+2))

trainpool = Pool(data=X, label=y, cat_features=[p, p+1], weight=w, feature_names=feature_names)
pool_no_label = cb.Pool(data=X, label=None,  cat_features=[p, p+1],  feature_names=feature_names)

model.fit(trainpool, eval_set=None, verbose_eval=500, plot=True)

print("score:{}, iter:{}, ntree:{}".format(model.get_best_score(), model.get_best_iteration(), model.tree_count_))
model.get_scale_and_bias()

shapvals = model.get_feature_importance(data=pool_no_label, prettified=True, type='ShapValues', shap_calc_type='Regular', reference_data=None, shap_mode='Auto')
explainer = shap.TreeExplainer(model)
shap_values = explainer.shap_values(pool_no_label)

```

**Comment by antipisa** [comment]
@ek-ak  thank you!

**Comment by andrey-khropov** [comment]
Fixed in b0e2cd83b5b3c4ecbcb16389c9ee979726461125
